Task groups guide (#205)
astronomer/airflow-guides: guides/task-groups.md

---
title: "Using Task Groups in Airflow"
description: "Using Task Groups to build modular workflows in Airflow."
date: 2021-05-05T00:00:00.000Z
slug: "task-groups"
heroImagePath: null
tags: ["DAGs", "Subdags", "Task Groups", "Best Practices"]
---

## Task Groups Overview

In December 2020, [Task Groups](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html#taskgroup) were introduced with the release of [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0). They are a UI grouping concept intended to replace SubDAGS.

> SubDAGs are DAGs embedded in DAGs whereas Task Groups exist in your original DAG. Tasks Groups allow you to group and organize your tasks within your DAG's graph view, without the added complexity and performance issues of SubDAGs (i.e. deadlocking, worker slots, etc.)

Reviewer: It's common in Overview sections to include a paragraph explaining what the guide will provide to the reader. You can see an example of it in [this guide](https://www.astronomer.io/guides/airflow-passing-data-between-tasks):

>There are a few methods you can use to implement data sharing between your Airflow tasks. In this guide, we'll walk through the two most commonly used methods, discuss when to use each, and show some example DAGs to demonstrate the implementation.

It would be great to have something similar here, just so readers know what to expect when diving in.

### Creating Task Groups

To use Task Groups you'll need to use the following import statement.

```python 

>     from airflow.utils.task_group import TaskGroup

Reviewer 2: ```suggestion
from airflow.utils.task_group import TaskGroup
```

```

When creating Task Groups, you can use the same dependency operators (<< and >>), these will be applied to all the tasks within the grouping.

```python
t0 = DummyOperator(task_id='start')

# Start Task Group definition
with TaskGroup(group_id='group1') as tg1:
    t1 = DummyOperator(task_id='task1')
    t2 = DummyOperator(task_id='task2')

    t1 >> t2
# End Task Group definition
    
t3 = DummyOperator(task_id='end')

t0 >> tg1 >> t3
```

This animated visual shows the above example with the UI interactions of expanding and collapsing the Task Group.

![UI Task Group](https://assets2.astronomer.io/main/guides/task-groups/task_groups_ui.gif)

> **Note:** When your task is within a task group, your callable task_id will be the task_id prefixed with the group_id (i.e. group_id.task_id). This ensures the uniqueness of the task_id across the DAG. This is important to remember when calling specific tasks like for XCOM passing or branching operator decisions. 

Reviewer 2: ```suggestion
**Note:** When your task is within a task group, your callable `task_id` will be the `task_id` prefixed with the `group_id` (i.e. `group_id.task_id`). This ensures the uniqueness of the task_id across the DAG. This is important to remember when calling specific tasks like for XCOM passing or branching operator decisions. 
```

> #### Dynamically Generating Task Groups

Author: I was thinking of it as Creating Task Groups && Overview as one level, and the rest as ways of creating or manipulating them. But, I think it also makes sense for them all to be the same level.

Task Groups can be dynamically generated to make use of patterns within your code. 

When calling an API, you may call a number of objects that have similar downstream tasks. This allows you to group by API objects for a simpler UI experience, that can be drilled into to see specific task statuses.

```python
for g_id in range(1,3):
    with TaskGroup(group_id=f'group{g_id}') as tg1:
        t1 = DummyOperator(task_id='task1')
        t2 = DummyOperator(task_id='task2')

        t1 >> t2

```

![Dynamic Task Group](https://assets2.astronomer.io/main/guides/task-groups/dynamic_task_groups.png)

##### Ordering Tasks Groups

> By default using a loop to generate your tasks groups, will put them in parallel. In cases where your tasks groups are dependent on elements of another task group then you'll want to run them sequentially. For example, when processing tables with foreign keys your primary table records will need to exist before you can load your foreign table.

Reviewer: I would add another paragraph explaining how the code below achieves this. Be explicit: "In the example below, we have a table with foreign keys, and we'd like to use task groups to process that table. (etc etc)"

```python
groups = []
for g_id in range(1,4):
    tg_id = f'group{g_id}'
    with TaskGroup(group_id=tg_id) as tg1:
        t1 = DummyOperator(task_id='task1')
        t2 = DummyOperator(task_id='task2')

        t1 >> t2

        if tg_id == 'group1':
            t3 = DummyOperator(task_id='task3')
            t1 >> t3
                
        groups.append(tg1)

[groups[0] , groups[1]] >> groups[2]
```

![Task Group Dependencies](https://assets2.astronomer.io/main/guides/task-groups/task_group_dependencies.png)

#### Nesting Task Groups

For additional complexity, you can nest Task Groups. 

```python
groups = []
for g_id in range(1,3):
    with TaskGroup(group_id=f'group{g_id}') as tg1:
        t1 = DummyOperator(task_id='task1')
        t2 = DummyOperator(task_id='task2')

        sub_groups = []
        for s_id in range(1,3):
            with TaskGroup(group_id=f'sub_group{s_id}') as tg2:
                st1 = DummyOperator(task_id='task1')
                st2 = DummyOperator(task_id='task2')

                st1 >> st2
                sub_groups.append(tg2)

        t1 >> sub_groups >> t2
        groups.append(tg1)

groups[0] >> groups[1]
```

> ![Nested Task Groups](https://assets2.astronomer.io/main/guides/task-groups/nested_task_groups.png)

Author: I don't know of any at the moment.

date: 2021-05-05T00:00:00.000Z
slug: "task-groups"
heroImagePath: null
tags: ["DAGs", "Subdags", "Task Groups", "Best Practices"]
---

## Overview

Prior to the release of [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0) in December 2020, in order to group tasks and create modular workflows within Airflow users had to use SubDAGs. SubDAGs were a way of presenting a cleaner looking DAG by capitalizing on code patterns. For example, ETL DAGs usually share a pattern of tasks that extract data from a source, transform the data, and load it somewhere. The SubDAG can visually group the repetitive tasks into one UI task, making the pattern between tasks clearer.

However, SubDAGs are really DAGs embedded in your DAG which can create both performance and functional issues:

- When a SubDAG is triggered, the SubDAG and child tasks take up worker slots until the entire SubDAG is complete. This can delay other task processing and, depending on your number of worker slots, can lead to deadlocking.
- SubDAGs have their own parameters, schedule, and enabled settings. When these are not consistent with their parent DAG, unexpected behavior can occur.

> Unlike SubDAGs, [Task Groups](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html#taskgroup) are just a UI grouping concept; allowing you to group and organize your tasks within your DAG's graph view without the added complexity and performance issues of SubDAGs (and with less code!).

Reviewer: It seems like the current usage of this term in the Airflow community uses capitalization, so I'm going to be consistent and capitalize Task Groups throughout the guide.

from airflow.utils.task_group import TaskGroup
```

In our first example, we will create a basic task group with two tasks inside and set its dependencies. You can use dependency operators (`<<` and `>>`) on task groups the same way that you can with individual tasks. Dependencies applied to the task group are applied across the inner tasks.

Below we instantiate our task group with our `with statement` and provide a `group_id`. Inside our task group, we define our two tasks and their dependencies to each other. Then we add our dependencies to the task group, which automatically applies the same dependencies across the inner tasks.  

```python
t0 = DummyOperator(task_id='start')

# Start Task Group definition
with TaskGroup(group_id='group1') as tg1:
    t1 = DummyOperator(task_id='task1')
    t2 = DummyOperator(task_id='task2')

    t1 >> t2
# End Task Group definition
    
t3 = DummyOperator(task_id='end')

# Set Task Group's (tg1) dependencies
t0 >> tg1 >> t3
```

This animated visual shows the above example with the UI interactions of expanding and collapsing the Task Group.

![UI Task Group](https://assets2.astronomer.io/main/guides/task-groups/task_groups_ui.gif)

> > **Note:** When your task is within a task group, your callable `task_id` will be the `task_id` prefixed with the `group_id` (i.e. `group_id.task_id`). This ensures the uniqueness of the task_id across the DAG. This is important to remember when calling specific tasks such as with XCOM passing or branching operator decisions.

Reviewer: ```suggestion
> **Note:** When your task is within a task group, your callable `task_id` will be the `task_id` prefixed with the `group_id` (i.e. `group_id.task_id`). This ensures the uniqueness of the task_id across the DAG. This is important to remember when calling specific tasks with XCOM passing or branching operator decisions.
```

## Dynamically Generating Task Groups

> Like tasks and [DAGs](https://www.astronomer.io/guides/dynamically-generating-dags), tasks groups can be dynamically generated to make use of patterns within your code. In an ETL DAG, you may call different API endpoints (i.e. a customer and contact table) that have similar downstream tasks but that can be processed independently. Here we can group the tasks by API endpoint, that like manually written task groups, can be drilled into from the Airflow UI to see specific tasks if needed. In the code below, we iterate over our desired groupings which creates multiple groups, with the same inner tasks where their passed parameters can change based on the `group_id`.

Reviewer: ```suggestion
[Just like with DAGs](https://www.astronomer.io/guides/dynamically-generating-dags), Task Groups can be dynamically generated to make use of patterns within your code. In an ETL DAG, you might have have similar downstream tasks that can be processed independently, such as when you need to call different API endpoints for data that ends up in the same table. For this use case, we can dynamically generate Task Groups by API endpoint. Just like manually written Task Groups, generated Task Groups can be drilled into from the Airflow UI to see specific tasks. 

In the code below, we use iteration to create multiple Task Groups. While the tasks and dependencies remain the same across Task Groups, we can change which parameters are passed in to each Task Group by iterating the `group_id`:
```
Tried to reword this to make it more clear based on what my understanding is here, feel free to modify as needed if I got any details wrong!

```python
for g_id in range(1,3):
    with TaskGroup(group_id=f'group{g_id}') as tg1:
        t1 = DummyOperator(task_id='task1')
        t2 = DummyOperator(task_id='task2')

        t1 >> t2
```

> This shows the expanded view of the task groups we generated above.

Reviewer: ```suggestion
This image shows the expanded view of the task groups we generated above:
```

![Dynamic Task Group](https://assets2.astronomer.io/main/guides/task-groups/dynamic_task_groups.png)

What if your task groups can't be processed independently? Next, we'll show how to call task groups and define dependencies between them.

## Ordering Tasks Groups

> By default, using a loop to generate your task groups will put them in parallel. If your tasks groups are dependent on elements of another task group, you'll want to run them sequentially. For example, when loading tables with foreign keys, your primary table records will need to exist before you can load your foreign table.

Reviewer: ```suggestion
By default, using a loop to generate your task groups will put them in parallel. If your tasks groups are dependent on elements of another task group, you'll want to run them sequentially. For example, when loading tables with foreign keys, your primary table records need to exist before you can load your foreign table.
```

> In the example below, our group 3 that we're dynamically generating has foreign keys to both group 1 and group 2 so we'll want to process it last. We create an empty list and append our task group airflow objects as they are generated. Using this list we reference the task groups and define their dependencies to each other.

Reviewer: ```suggestion
In the example below, our third dynamically generated Task Group requires foreign keys from both our first and second dynamically generated Task Groups, so we'll want to process it last. To do this, we'll create an empty list and append our Task Group objects as they are generated. Using this list, we can reference the Task Groups and define their dependencies to each other:
```

```python
groups = []
for g_id in range(1,4):
    tg_id = f'group{g_id}'
    with TaskGroup(group_id=tg_id) as tg1:
        t1 = DummyOperator(task_id='task1')
        t2 = DummyOperator(task_id='task2')

        t1 >> t2

        if tg_id == 'group1':
            t3 = DummyOperator(task_id='task3')
            t1 >> t3
                
        groups.append(tg1)

[groups[0] , groups[1]] >> groups[2]
```

> This shows the task group dependencies. 

Reviewer: ```suggestion
The following screenshot shows how these Task Groups appear in the Airflow UI:
```

![Task Group Dependencies](https://assets2.astronomer.io/main/guides/task-groups/task_group_dependencies.png)

> ## Conditioning on Task Groups

Reviewer: ```suggestion
### Conditioning on Task Groups
```
Sorry, I know I said to make these all `h2` but I didn't realize this was dependent on the previous example 😬

Author: That makes sense, yeah I added in this section after the last review when I realized I did a thing in my example without ever addressing it.

> In the above example, we added an additional task to group 1 based on our `group_id`. This was to demonstrate that even though we're dynamically creating task groups and tasks to take advantage of patterns we can still introduce variations to the pattern while avoiding code redundancies from building each task group definition manually.

Reviewer: ```suggestion
In the above example, we added an additional task to `group1` based on our `group_id`. This was to demonstrate that even though we're dynamically creating Task Groups to take advantage of patterns, we can still introduce variations to the pattern while avoiding code redundancies from building each task group definition manually.
```

## Nesting Task Groups

> For additional complexity, you can nest task groups. Building on our previous ETL example, when calling API endpoints we may need to process new records for each endpoint before we can process updates to them.

Reviewer: ```suggestion
For additional complexity, you can nest Task Groups. Building on our previous ETL example, when calling API endpoints we may need to process new records for each endpoint before we can process updates to them.
```

> In the code below, our top level task groups represents our new and updated record processing, while the nested task group represents our API endpoint processing. 

Reviewer: ```suggestion
In the following code, our top-level Task Groups represent our new and updated record processing, while the nested Task Groups represent our API endpoint processing:
```

```python
groups = []
for g_id in range(1,3):
    with TaskGroup(group_id=f'group{g_id}') as tg1:
        t1 = DummyOperator(task_id='task1')
        t2 = DummyOperator(task_id='task2')

        sub_groups = []
        for s_id in range(1,3):
            with TaskGroup(group_id=f'sub_group{s_id}') as tg2:
                st1 = DummyOperator(task_id='task1')
                st2 = DummyOperator(task_id='task2')

                st1 >> st2
                sub_groups.append(tg2)

        t1 >> sub_groups >> t2
        groups.append(tg1)

groups[0] >> groups[1]
```

> This shows the expanded view of the nested task groups we generated above.

Reviewer: ```suggestion
The following screenshot shows the expanded view of the nested Task Groups in the Airflow UI:
```

![Nested Task Groups](https://assets2.astronomer.io/main/guides/task-groups/nested_task_groups.png)

## Takeaways

> Task groups are a dynamic and scalable UI grouping concept that solve for the functional and performance issues of SubDAGs. 

Reviewer: ```suggestion
Task Groups are a dynamic and scalable UI grouping concept that eliminate the functional and performance issues of SubDAGs. 
```

> Ultimately task groups give you the flexibility to group and organize your tasks in a number of ways, to help guide your usage of them think about:

Reviewer: ```suggestion
Ultimately, Task Groups give you the flexibility to group and organize your tasks in a number of ways. To help guide your implementation of Task Groups, think about:
```

> - What patterns exist in your DAG?

Reviewer: ```suggestion
- What patterns exist in your DAGs?
```

# Start Task Group definition
with TaskGroup(group_id='group1') as tg1:
    t1 = DummyOperator(task_id='task1')
    t2 = DummyOperator(task_id='task2')

    t1 >> t2
# End Task Group definition
    
t3 = DummyOperator(task_id='end')

# Set Task Group's (tg1) dependencies
t0 >> tg1 >> t3
```

In the Airflow UI, Task Groups look like tasks with blue shading. When we expand `group1` by clicking on it, we see blue circles where the Task Group's dependencies have been applied to the grouped tasks. The task(s) immediately to the right of the first blue circle (`t1`) get the group's upstream dependencies and the task(s) immediately to the left (`t2`) of the last blue circle get the group's downstream dependencies. 

![UI Task Group](https://assets2.astronomer.io/main/guides/task-groups/task_groups_ui.gif)

> **Note:** When your task is within a Task Group, your callable `task_id` will be the `task_id` prefixed with the `group_id` (i.e. `group_id.task_id`). This ensures the uniqueness of the task_id across the DAG. This is important to remember when calling specific tasks with XCOM passing or branching operator decisions.

## Dynamically Generating Task Groups

> [Just like with DAGs](https://www.astronomer.io/guides/dynamically-generating-dags), Task Groups can be dynamically generated to make use of patterns within your code. In an ETL DAG, you might have have similar downstream tasks that can be processed independently, such as when you call different API endpoints for data that needs to be processed and stored in the same way. For this use case, we can dynamically generate Task Groups by API endpoint. Just like with manually written Task Groups, generated Task Groups can be drilled into from the Airflow UI to see specific tasks. 

Reviewer: ```suggestion
[Just like with DAGs](https://www.astronomer.io/guides/dynamically-generating-dags), Task Groups can be dynamically generated to make use of patterns within your code. In an ETL DAG, you might have similar downstream tasks that can be processed independently, such as when you call different API endpoints for data that needs to be processed and stored in the same way. For this use case, we can dynamically generate Task Groups by API endpoint. Just like with manually written Task Groups, generated Task Groups can be drilled into from the Airflow UI to see specific tasks. 
```
